compiler/rustc_lint/src/redundant_prelude_imports.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{lints::RedundantPreludeImportsDiag, EarlyContext, EarlyLintPass, LintContext};
+use rustc_ast as ast;
+use rustc_span::symbol::{sym, Symbol};
+use rustc_span::Span;
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use std::rc::Rc;
+use std::string::ToString;
+
+declare_lint! {
+    /// The `redundant_prelude_imports` lint detects unnecessary imports in
+    /// std::prelude that can be eliminated.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// # #![warn(redundant_prelude_imports)]
+    /// use std::mem::drop;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// unnecessary prelude import  'std::mem::drop'
+    pub(super) REDUNDANT_PRELUDE_IMPORTS,
+    Allow,
+    "detects unnecessary imports in std::prelude that can be eliminated"
+}
+
+declare_lint_pass!(RedundantPreludeImports => [REDUNDANT_PRELUDE_IMPORTS]);
+
+#[derive(Debug)]
+struct RedundantImport<'a, 'b> {
+    item: &'a ast::Item,
+    prelude_imports: &'b BTreeMap<&'b str, Symbol>,
+    remove_imports: Vec<(Span, Vec<String>)>,
+    reserve_imports: Vec<(Span, Vec<String>)>,
+}
+
+impl EarlyLintPass for RedundantPreludeImports {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &ast::Item) {
+        if let ast::ItemKind::Use(ref use_tree) = item.kind {
+            // Use in std::prelude
+            let prelude_imports: BTreeMap<&str, Symbol> = BTreeMap::from([
+                ("std::marker::Copy", sym::rust_2015),
```

Comment:
Comparing strings is super fragile.

Further, personally speaking I'd prefer if the list of prelude items wasn't hard-coded but computed once dynamically. This way, we can easily extend the prelude in future editions (like Rust 2024) without having to worry about this lint becoming outdated.

Reply:
> Comparing strings is super fragile.
> 
> Further, personally speaking I'd prefer if the list of prelude items wasn't hard-coded but computed once dynamically. This way, we can easily extend the prelude in future editions (like Rust 2024) without having to worry about this lint becoming outdated.


Thank you very much. Totally agree. 
But since I am a newbie, I am not sure at which stage and in what way the prelude list is calculated? And in which struct stored it? Could you give me some advice?